### ISSUE TITLE
No common gene when running Tangram in Sopa-CLI model

### ISSUE THREAD

**Issue by KunHHE** [gustaveroussy/sopa#174]
Hi @quentinblampey, I used CLI mode to run sopa, and want to use Tangram directly with sopa for my merfish data. But it error: no common gene found between .zarr and .h5ad reference. Looks like in the .h5ad reference, it hides gene names and gene ensemble id jump out for the cell type training, that is why two datasets cannot match. 
Because in the jupyter I run :
gene_name_mapping = adata_sc.var['feature_name'] 
adata_sc.var_names = gene_name_mapping

Then the overlapped genes showed up for training. 

Is there any way in the CLI running to figure out?

Thansk!

**Comment by quentinblampey** [comment]
Hi @KunHHE, indeed `adata_sc.var_names` should contain the gene names.
Could you update your reference (as you did), and update your .h5ad file?

> NB: I believe it's easier to update the reference rather than adding an argument to the CLI

**Comment by KunHHE** [comment]
Thanks very much! @quentinblampey So you mean I update this using the code for the reference.h5ad: gene_name_mapping = adata_sc.var['feature_name']; adata_sc.var_names = gene_name_mapping. Then save it and reuse it in sopa for Tangram?

**Comment by quentinblampey** [comment]
Yes @KunHHE, exactly! Let me know if this works

**Comment by KunHHE** [comment]
HI, @quentinblampey, tested and it works. But I have a question, single cell resolution-like technologies like merscope, Xenium, Visium HD are recommended using uniform mode? based on the introduction from Tangram github. non-single cell like technologies are recommended using rna_count_based density_prior. In the CLI of sopa, it's not flexible switching to 'uniform'?
So after the Tangram is done, it will automatically adding predicted cell_type to the .obs in the zarr-Table?

Thanks!!! 

(sopa) C:\Users\hekun>sopa annotate tangram C:/Users/hekun/Downloads/S3R1.zarr --sc-reference-path C:/Users/hekun/Downloads/M1_modified.h5ad --cell-type-key cell_type
C:\Users\hekun\miniconda3\envs\sopa\lib\site-packages\dask\dataframe\__init__.py:31: FutureWarning: The legacy Dask DataFrame implementation is deprecated and will be removed in a future version. Set the configuration option `dataframe.query-planning` to `True` or None to enable the new Dask Dataframe implementation and silence this warning.
  warnings.warn(
C:\Users\hekun\miniconda3\envs\sopa\lib\site-packages\anndata\utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.
  warnings.warn(msg, FutureWarning)
[INFO] (sopa.annotation.tangram.run) Using device: cpu
[INFO] (sopa.annotation.tangram.run) Running on level 0
[INFO] (sopa.annotation.tangram.run) Subsampling reference to 10000 cells...
[INFO] (sopa.annotation.tangram.run) (n_obs_spatial=22373, n_obs_ref=10000)
[INFO] (sopa.annotation.tangram.run) --- Split 1 / 3 ---
[INFO] (sopa.annotation.tangram.run) Using raw counts for the spatial adata object
[INFO] (sopa.annotation.tangram.run) Genes with zero counts: 0 spatial, 3312 ref
[INFO] (sopa.annotation.tangram.run) Keeping 404 shared genes
INFO:root:Allocate tensors for mapping.
INFO:root:Begin training with 404 genes and rna_count_based density_prior in cells mode...
INFO:root:Printing scores every 100 epochs.
Score: 0.189, KL reg: 0.322

**Comment by quentinblampey** [comment]
I'm not sure to understand your question. We use a uniform density, indeed. 
Regarding the predictions, tangram saves them under `tangram_ct_pred`, but we then move it under `tangram_pred`, so you're looking at the right thing

**Comment by KunHHE** [comment]
Thanks very much@quentinblampey. Can I ask you if we have multiple samples for merscope or xenium data, let's say 30 samples, then we will have to run Tangram one-by-one, because I like sopa-Tangram here, as it splits the Tangram run, so the RAM memory won't be filled full.

**Comment by quentinblampey** [comment]
Yes @KunHHE by default we run this sample by sample, but you could also decide to concatenate your samples into one AnnData object, and then run Tangram on this concatenated object. Be careful to the batch effect though: if you have a lot of batch effect, this could affect the annotation after concatenation

**Comment by KunHHE** [comment]
> I'm not sure to understand your question. We use a uniform density, indeed. Regarding the predictions, tangram saves them under `tangram_ct_pred`, but we then move it under `tangram_pred`, so you're looking at the right thing

Hi @quentinblampey, sorry I am looping back to this question. My question is as you mentioned above the tamgram annotation here used "uniform" density.  But in the CLI mode, I can see "INFO:root:Begin training with 270 genes and rna_count_based density_prior in cells mode." So it uses "rna_count_based density", do you think the API can make an argue on this? Thank you!

**Comment by quentinblampey** [comment]
Hi @KunHHE, you're correct. I was computing the uniform densities, but then still using the default mode in Tangram; my bad.

I corrected this, now it will use the uniform mode by default, and I added the `density_prior` argument so that we can change it if desired. It will be released in `sopa==2.0.3`
